gtk: implement bell

pluiedev opened the pull request:
This PR implements a more lightweight alternative to #5326 that contains features that I personally think Just Make Sense for the bell.

No configs, no GStreamer stuff, just sane defaults to get us started.

pluiedev committed:
gtk(bell): use `gdk.Surface.beep` for bell

Co-authored-by: Jeffrey C. Ollie <jeff@ocjtech.us>

pluiedev committed:
gtk: implement bell

Co-authored-by: Jeffrey C. Ollie <jeff@ocjtech.us>

pluiedev committed:
gtk(bell): mark tab as needing attention on bell

pluiedev commented:
The system bell isn't exactly working on NixOS + KDE at the moment, since GTK 4.16 doesn't yet support the *new* system bell protocol which KWin implements. You'd have to ask a GNOME user what it exactly does ;)

As for the tab indicator thing:
![image](https://github.com/user-attachments/assets/827918f0-2f7b-4893-9ea7-1c121fcbb18e)

tristan957 commented:
The UI change looks good.

tristan957 commented:
Should you add `Co-authored-by: jcollie...`?

pluiedev commented:
As I've already said in https://github.com/ghostty-org/ghostty/discussions/2710#discussioncomment-12017099, I personally feel like offering too many options falls in the same trap as other mega-customizable settings, in that the implementation costs don't outweigh the benefits of having that many degrees of freedom. Besides, this is intentionally designed to allow for more bell settings in the future, as you could just add if statements before the labelled blocks. I just think that this is a sensible default that works for most people.

About GStreamer... the fact of the matter is that we have to add more bindings, add deps to Nix (and Flatpaks, Snaps, etc.), embed our own audio files, and so on... I don't think that's worth it, not for me personally. Then again, this could be part of a future addition. I'd just like to see *some* kind of bell feature implemented inside Ghostty at the moment, which is evidently important to quite a few people.

tristan957 commented:
> I think that the lack of configuration, even just the ability to enable/disable the bell, is not a good choice.

I agree with this. Adding a `bell` and `bell-features` config options would be good, which I think @jcollie had in his original PR.

Adding the custom sound could come in a separate PR based on jcollie's work.

tristan957 commented:
> I personally feel like offering too many options falls in the same trap as other mega-customizable settings, in that the implementation costs don't outweigh the benefits of having that many degrees of freedom

This ship has sailed.

pluiedev commented:
Fine - I think enabling/disabling the system bell function makes sense, but I'm very much against implementing GStreamer stuff in this PR. It simply rose too many questions in the last PR and impeded the ball from actually rolling

pluiedev committed:
gtk(bell): add `bell-features` config option

Co-authored-by: Jeffrey C. Ollie <jeff@ocjtech.us>

jcollie commented:
My latest patch does not embed any audio files. I agree that my initial patches that did that were a mistake. If someone wants to use a custom audio file they would need to supply it themselves. 

Setting attention on the tab is unobtrusive enough to leave always on I think.

In Gnome, the system beep is semi-configurable from a list of options. You can see what they sound like by clicking on the options.

![image](https://github.com/user-attachments/assets/ee88e3d2-0e59-4af9-adfa-b585e34a4020)

tristan957 commented:
> Fine - I think enabling/disabling the system bell function makes sense (the tab thing is such a small visual fluorish I'm not really sure if we should make it configurable either)

Agree completely here.

> I'm very much against implementing GStreamer stuff in this PR. It simply rose too many questions in the last PR and impeded the ball from actually rolling

Yes, let's land the basics, and then @jcollie can rebase on it. I am +1 on custom sounds :smile:

pluiedev commented:
Added `bell-features` and also added @jcollie to a bunch of these commits as a co-author 'cos he's been part of this since even the first prototype, and this feels more right :p

mitchellh reviewed (changes requested):
One very simple change request.

The rest looks very good. Thanks for paring down the PR, this is very easy and much more palatable to merge. 😄 The config types/options may need to be tuned some more before a release but it's certainly good enough for a tip build for now.

Aside:

> I personally feel like offering too many options falls in the same trap as other mega-customizable settings, in that the implementation costs don't outweigh the benefits of having that many degrees of freedom.

I understand this opinion, but that isn't this project. Ghostty's config philosophy is "zero-config defaults" (e.g. 90+% of people don't need to touch a single thing) with extreme flexibility otherwise, I've said this publicly before that I want Ghostty to be one of the most customizable terminals out there _for people who want it_. So, we'll have a lot of config options over time (we already do).

pluiedev commented on src/config/Config.zig:
```diff
@@ -5669,6 +5685,11 @@ pub const AppNotifications = packed struct {
 
+/// See bell-features
+pub const BellFeatures = packed struct {
+    system: bool = true,
```
It can do more than just audio beeps - on KDE at least supposedly it also does visual flashes. You're right that `system` is *kind of* a bad name though.

tristan957 reviewed (approved):
Thanks to you and @jcollie for getting the first iteration of this done!

tristan957 commented:
Maybe a force push would kick github runners to try again?

mitchellh commented:
I'm willing to risk it. LOL. CI on main.

ncfavier commented on src/config/Config.zig:
```diff
@@ -1848,6 +1848,22 @@ keybind: Keybinds = .{},
+///
+/// Valid values are:
+///
+///  * `system` (default)
```
It says "default" here but the value defaults to `false`.

00-kat replied:
Oh you had commented this before.  I randomly noticed this a few minutes ago too and created a pull requested → #7148.

ncfavier commented:
This doesn't seem to set the X11 urgency hint, which is the only thing I care about the bell doing. Would it be possible to support that?

jcollie commented:
> This doesn't seem to set the X11 urgency hint, which is the only thing I care about the bell doing. Would it be possible to support that?

It's possible, but please open a new feature discussion to discuss it.

00-kat commented:
In my opinion, create a new Discussion instead of bumping the old one.

ncfavier commented:
Opened https://github.com/ghostty-org/ghostty/discussions/7124
